man/man1/snapmass.1
.TH SNAPMASS 1NEMO "14 August 2020"
.SH NAME
snapmass \- add or modify masses in a snapshot
.SH SYNOPSIS
\fBsnapmass in=\fPsnap_in \fBout=\fPsnap_out [parameters=values...]
.SH DESCRIPTION
\fIsnapmass\fP adds or modifies the masses in a snapshot. 
The keyword \fBmass\fP
can be used to give a \fIbodytrans(3NEMO)\fP expression. 
.PP
Alternatively masses can be copied from another input snapshot.
.SH PARAMETERS
The following parameters are reckognized in any order if the keyword is
also given:
.TP 25
\fBin=\fIsnap_in\fP
Input data is read from \fIsnap_in\fP, which must be in standard
\fIsnapshot(5NEMO)\fI format [no default].
.TP
\fBout=\fIsnap_out\fP
Copied output data is written to \fIsnap_out\fP, which will also be in 
\fIsnapshot(5NEMO)\fP format [No default].
.TP
\fBmass=\fIexpression\fP
Expression for the new masses, using \fIbodytrans(3NEMO)\fP to parse
the expression. 
If nothing supplied, either 
\fBinmass=\fP or \fBmassname=\fP will have to be used.
[default: none].
.TP
\fBinmass=\fImass-file\fP
If used, the masses from this snapshot file are copied to that of the
\fIsnap_out\fP file. [default: none].
.TP
\fBmassname=\fIname\fP
Identification of the mass spectrum. It is normally only a function
of mass, so \fBn(m)\fP will suffice. If nothing specified, all
stars will have the same mass (total mass M=1).
[default: none].
.TP
\fBmassexpr=\fIexpr\fP
Expression for the mass-spectrum. In addition to the mass \fIm\fP from
the \fBmassname\fP it may contain any variables contained in the 
\fBmasspars\fP to be specified next. [default: \fBpow(m,p)\fP].
.TP
\fBmasspars=\fIpars\fP
List of parameters and their values. They can be used by name in the
\fBmassexpr\fP specified before. [default: \fBp,0.0\fP].
.TP
\fBmassrange=\fImlow:mhigh\fP
Valid massrange. A lower and higher cutoff must be supplied. The massrange
has arbitrary units, since all masses will be rescaled to set the total
mass M=1. [default: \fB1,1\fP].
.TP
\fBseed=\fP\fIseed\fP
seed for the random number  generator (default: a value 0, which will
be  converted  into  a  unique  new
value  using  UNIX's clock time, in 
seconds since  once-upon-a-time-in-the-seventies).
[Default: 0].
.TP
\fBccd=\fP
Input CCD image with mapvalues that represent mass. By default not used.
.TP
\fBnorm=\fP\fItotal_mass\fP
New total mass required for normalization. By default the masses are not
scaled. See also \fIsnapscale(1NEMO)\fP.
[Default: not used].
.TP
\fBaux=t|f\fP
Store the "mass" in the Aux slot of a \fIsnapshot(5NEMO)\fP. [false]
.SH SEE ALSO
snapmstat(1NEMO), snapscale(1NEMO), bodytrans(3NEMO)
.SH AUTHOR 
Peter Teuben
.SH FILES
.nf
.ta +3.0i
~/src/nbody/trans	snapmass.c
.fi
.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.5i
1-Aug-88	V1.0: created             	PJT
5-dec-88	V1.1: mass keyword added	PJT
5-feb-89	V1.2: bodytrans() used     	PJT
13-may-91	V1.3: helpvec + doc added	PJT
29-may-97	V2.0: mass{name,expr,pars,range} from mkplummer 	PJT
30-jul-97	V2.1: added norm=           	PJT
14-aug-2020	V2.3: add aux=		PJT
.fi


src/nbody/trans/snapmass.c
/*
 *  SNAPMASS:  set masses in a snapshot
 *
 *	1-Aug-88	V1.0 created      				PJT
 *	5-dec-88	V1.1 allow all masses same through one keyword  PJT
 *				(proposed - to be coded )
 *	5-feb-89	V1.2 allow bodytrans(3NEMO) expression for mass PJT
 *	13-may-91	V1.3 added helpvec etc.				PJT
 *      23-nov-93       V1.3a cosmetic
 *      21-dec-95       V1.4 process all snapshots
 *	28-may-97	2.0 allow bodytrans and generic function	PJT
 *				(code taken from mkplummer)
 *      24-jul-97       2.1 added norm=
 *	 8-sep-01       a   init_xrandom
 *      29-aug-06       c   prototypes for frandom() and getrfunc() properly used
 *      12-jul-2019     2.2 added ccd=
 *      14-aug-2020     2.3 allow Aux to be used
 *
 */
#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>
#include <image.h>
#include <loadobj.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
#include <snapshot/get_snap.c>
#include <snapshot/put_snap.c>

#include <bodytrans.h>


string defv[] = {
    "in=???\n		      input (snapshot) file",
    "out=???\n                output (snapshot) file",
    "mass=\n		      expression for new masses",
    "inmass=\n                extra file, if to use their masses only",
    "massname=\n              If used,  Mass-function name (mostly n(m))",
    "massexpr=pow(m,p)\n      Mass function expression (e.g. pow(m,p))",
    "masspars=p,0.0\n         Mass function parameters (e.g. p,0.0)",
    "massrange=1,1\n          Range for mass-spectrum (e.g. 1,2)",
    "seed=0\n                 Random seed",
    "ccd=\n                   Input CCD with mapvalues that represent mass",
    "norm=\n                  Normalization value for the total mass (if used)",
    "aux=f\n                  Store in Aux instead?",
    "VERSION=2.3\n            14-aug-2020 PJT",
    NULL,
};

string usage="(re)assign masses to a snapshot";

string cvsid="$Id$";

#define TIMEFUZZ	0.0001	/* tolerance in time comparisons */

extern real_proc getrfunc(string , string , string , int *);

void nemo_main(void)
{
    stream instr, inmassstr, outstr, ccdstr;
    real   tsnap, tsnapmass, mass_star, mtot, mrange[2], norm;
    real   xpos, ypos, xmin, ymin, idx, idy;
    Body  *btab = NULL, *bp;
    Body  *bmasstab = NULL, *bmassp;
    imageptr iptr = NULL;
    int i, n, nbody, nbodymass, bits, bitsmass, seed, ix, iy;
    rproc_body  bfunc;
    real_proc   mfunc;
    bool  Qnorm, first = TRUE;
    bool Qaux = getbparam("aux");
    string what = (Qaux ? strdup("aux") : strdup("mass"));

    instr = stropen(getparam("in"), "r");
    if (hasvalue("mass")) {
        dprintf(0,"Using a bodytrans %s expression\n", what);
        bfunc = btrtrans(getparam("mass"));     /* use bodytrans expression */
        mfunc = NULL;
        inmassstr = NULL;
    } else if (hasvalue("massname")) {
        dprintf(0,"Using a general functional mass expression\n");
        mysymbols(getargv0());
        n=1;
        mfunc = getrfunc(getparam("massname"),
                         getparam("massexpr"),getparam("masspars"),&n);
        if (nemoinpr(getparam("massrange"),mrange,2)!=2)
            error("Need two numbers in massrange=");
        dprintf(1,"Massrange from %f : %f\n",mrange[0],mrange[1]);
        bfunc = NULL;
        inmassstr = NULL;
    } else if (hasvalue("inmass")) {
        inmassstr = stropen(getparam("inmass"),"r");
    } else if (hasvalue("ccd")) {
      warning("new ccd mode");
      ccdstr = stropen(getparam("ccd"),"r");
      read_image(ccdstr,&iptr);
      idx = 1.0/Dx(iptr);
      idy = 1.0/Dy(iptr);
      xmin = Xmin(iptr);
      ymin = Ymin(iptr);
      inmassstr = NULL;
      mfunc = NULL;
      bfunc = NULL;      
    } else 
    	error("One of: mass=, massname=, inmass=, ccd= must be given");
    outstr = stropen(getparam("out"), "w");
    seed = init_xrandom(getparam("seed"));
    Qnorm = hasvalue("norm");
    if (Qnorm) norm = getdparam("norm");


    for(;;) {
					/* input main data */
    	get_history(instr);                    /* skip history & comments */
        if (!get_tag_ok(instr, SnapShotTag)) {
            break;
            error("Snapmass (in): Need a snapshot");
        }
        get_snap(instr, &btab, &nbody, &tsnap, &bits);
        if (bits&MassBit) {
            dprintf(0,"Warning: existing masses overwritten\n");
            if (!Qnorm) {
            	Qnorm = TRUE;
        	for (bp=btab, i=0, norm=0.0; i<nbody; bp++,i++)            
            	    norm += Mass(bp);
            }
        }
        if (!(bits&TimeBit)) {
            dprintf(1,"Warning: time=0.0 assumed\n");
            tsnap = 0.0;
        }
        
        if (inmassstr) { 		/* input mass data from a file */
    	    get_history(inmassstr);
            if (!get_tag_ok(inmassstr, SnapShotTag))
       		error("Snapmass (inmass): Need a snapshot");
            get_snap(inmassstr, &bmasstab, &nbodymass, &tsnapmass, &bitsmass);
            if (nbodymass != nbody) {
	        if (nbodymass < nbody)
                    error("too few bodies (%d < %d)",nbodymass,nbody);
                else
                    warning("too many bodies (%d > %d)",nbodymass,nbody);
            }
            if ((bitsmass & MassBit) == 0) {
  	        // unlink(getparam("out"));
                error("essential mass data missing\tbits = %x\n", bits);
            }
        }
        				/* output */
        if (first) {
            put_history(outstr);
            first = FALSE;
        }
        if (inmassstr) {
	    if (nbody==nbodymass)
	        for (bp=btab, bmassp=bmasstab; bp<btab+nbody; bp++, bmassp++)
		  if (Qaux)
		    Aux(bp) = Mass(bmassp);
		  else
		    Mass(bp) = Mass(bmassp);
	    else
	        for (bp=btab, bmassp=bmasstab; bp<btab+nbody; bp++)
		  if (Qaux)
	            Aux(bp) = Mass(bmassp);
		  else
	            Mass(bp) = Mass(bmassp);
        } else if (bfunc) {
            for (bp=btab, i=0; i<nbody; bp++,i++)
	      if (Qaux)
                Aux(bp) = bfunc(bp, tsnap, i);
	      else
                Mass(bp) = bfunc(bp, tsnap, i);
        } else if (mfunc) {
            for (bp=btab, i=0; i<nbody; bp++,i++)
	      if (Qaux)
	        Aux(bp) = frandom(mrange[0], mrange[1], mfunc);
	      else
	        Mass(bp) = frandom(mrange[0], mrange[1], mfunc);
	} else if (iptr) {
	    for (bp=btab, i=0; i<nbody; bp++,i++) {
	      xpos = (Pos(bp)[0]-xmin)*idx + 0.5;    /* fractional cell index   0..nx */
	      ypos = (Pos(bp)[1]-ymin)*idy + 0.5;
	      ix = (int) floor(xpos);            /* cell index:   0 .. nx-1 */
	      iy = (int) floor(ypos);
	      if (ix<0 || iy<0 || ix>=Nx(iptr) || iy>=Ny(iptr))
		if (Qaux)
      		  Aux(bp) = 0.0;
		else
      		  Mass(bp) = 0.0;
	      else
		if (Qaux)
		  Aux(bp) = MapValue(iptr,ix,iy);
		else
		  Mass(bp) = MapValue(iptr,ix,iy);
	    }
        } else             
            error("bad flow logic");

        for (bp=btab, i=0, mtot=0.0; i<nbody; bp++,i++)            
	  mtot += (Qaux ? Aux(bp) : Mass(bp));
        dprintf(0,"Total %s: %g\n", what , mtot);
        if (Qnorm) {
	  dprintf(0,"Rescaling total %s to %g\n", what , norm);
            for (bp=btab, i=0; i<nbody; bp++,i++)
	      if (Qaux)
                Aux(bp) *= norm/mtot;
	      else
                Mass(bp) *= norm/mtot;
        }
	bits |= MassBit;    /* turn mass bit on anyways */
	if (Qaux)
	  bits |= AuxBit;
        put_snap(outstr, &btab, &nbody, &tsnap, &bits);
    }

    strclose(instr);
    if (inmassstr) strclose(inmassstr);
    strclose(outstr);
}
